### DISCUSSION TITLE
How to use LearnLib in a practical system？

### DISCUSSION THREAD

**Discussion by tao2years** [LearnLib/learnlib#97] [Q&A] [answered]
I am currently exploring the field of automata learning and its practical applications. I have experimented with the examples provided by LearnLib, such as the "cav2015 CoffeeMachine" example. However, I noticed that these examples assume a predetermined set of inputs and outputs defined by alphabets. In real-world systems, the inputs often involve parameters and value domains that require additional processing.

After delving into the topic further, including studying SSFLA 2011 Automata Learning from Theory to Application, I am now seeking guidance on how to approach learning automata for such real-world systems. Specifically, I have the following questions:

(1) Are there any practical examples available that demonstrate learning automata for systems with diverse inputs, taking into account the variability of inputs?

(2) If there are no specific examples available, what algorithms should I consider using? How should I handle the diverse inputs and outputs to construct the automata? Additionally, how can I determine the completeness of the generated automata, such as identifying equivalent classes?

I would greatly appreciate any insights, resources, or guidance you can provide to help me address these questions and navigate the challenges of learning automata for real-world systems.

Thank you for your time and assistance.

**Comment by mtf90** [comment, marked as answer]
Dear @tao2years, this problem is not specific to automata learning but can be found in other model-based techniques such as (model-) verification and (model-based) testing, too. So mabye, if you broaden your field of research a little bit, you are able to come across possible solutions that are suitable for you. While the definition of input symbols can be seen as an additional challenge, it is also a powerful tool to steer the degree of abstraction (which is one of the main appeals of using model-based techniques in the first place, in my opinion).

As for the learning process, two approaches come to (my) mind:
* [Automated Alphabet Abstraction Refinement (AAAR)](https://doi.org/10.1007/978-3-642-18275-4_19) allows you to deal with infinite input domains (such as integer numbers) by using only a single representative for the (finite) amount of equivalently behaving partition classes of the domain (e.g., -4 for all negative numbers, +3 for all positive numbers). The nice thing about this approach is that the discovery of these equivalence classes can be formulated as part of the learning process as well so that it takes place automatically. The current (development) version of LearnLib provides two implementations for this AAAR principle and the [examples module](https://github.com/LearnLib/learnlib/tree/develop/examples/src/main/java/de/learnlib/example/aaar) contains some code to get your feet wet.
* [Register automata](https://doi.org/10.1007/978-3-319-10431-7_18) upgrade the expressiveness of the inferred models to integrate data-flow (via registers) as a first-class citizen. Currently, there exists [ralib](https://github.com/LearnLib/ralib) as an extension of LearnLib to support this type of automaton and the inference thereof. While the developers seem to be busy in the last couple of weeks, I'm currently not involved that much in the project, so I cannot really comment on its current state. Maybe @fhowar wants to chime in on this if you have further questions.

As for the (successful) application of automata learning in practical scenarios, you may find some further examples [here](https://doi.org/10.1109/ICSTW.2013.60), [here](https://doi.org/10.1007/978-3-642-16558-0_54), [here](https://doi.org/10.1007/978-3-319-67113-0_12), [here](https://doi.org/10.1007/978-3-030-90870-6_28), or [here](https://doi.org/10.1007/978-3-319-25423-4_5).


I hope this gives you a little nudge in the right direction to solve your problems.

**Reply by tao2years** [reply]
Thanks for your patient relay! 
I will try these methods and materials you mentioned above.

**Reply by mtf90** [reply]
You're welcome. If you have any further questions, feel free to ask. Otherwise, you may mark it as done/answered for others to see.
